Add validations

lrharris215 opened the pull request:
I added a MoveValidator that uses the Validator interface. The MoveValidator checks if the player input is a valid move, (between 1 and 7, and whether the column is full). There is also an error message displayed if the move is invalid.

lrharris215 committed:
Add: Validator interface and MoveValidator

The MoveValidator checks if the users input is a valid move.
If the number is > 7 or < 1, it is invalid. The validator
also checks if the column chosen has empty spots.

lrharris215 committed:
Test: Add MoveValidator tests

Add tests for the MoveValidator. Inputs for isValid must be
a number between 1 and 7(inclusive), and the spot on the
board must be empty.

lrharris215 committed:
Add: MoveValidator to the Game

Implemented the MoveValidator by adding it to the game class.
Players now take in the validator as an argument during
takeTurn. Also created the TestValidator to stub in for testing.

lrharris215 committed:
Refactor: Moved Validators and Scanner out of Game

Moved the PlayerScanner and MoveValidator to args for Player
instead of Game. Only the Player uses those two classes.

lrharris215 committed:
Add: Error Message for failed MoveValidation

Added an Error function to the Presenter. The program now
lets the player know when they have submitted an invalid move.

lrharris215 committed:
Refactor: Made Disc's symbol unchangeable

Since all discs use the same symbol, I removed the argument
from the constructor and just set it to what it should be.

kudukacar commented on src/main/java/com/connectfour/Player.java:
```diff
@@ -3,19 +3,27 @@
+        }
         colChoice = colChoice - 1;
         placeDisc(board, colChoice);
     }
```
Interesting the player knows to reduce the colChoice by 1 and also places the disc.  Is this the responsibility of the player?  I would think the board would adjust the colChoice as needed - it's an implementation detail of the board.  And placing the disc is an orchestration of the game.

lrharris215 replied:
Moved the -1 from Player to Board. Also removed the placeDisc function from Player, because I can just call Board#addDisc inside Player#takeTurn instead.

kudukacar replied:
board.addDisc really should not be the player's responsibility.  Should be the game's responsibility.  The player just provides it's move.  The game updates the board to reflect the player's move.

lrharris215 replied:
I see what you mean. Fixed.

kudukacar commented on src/test/java/com/connectfour/MoveValidatorTest.java:
```diff
@@ -0,0 +1,49 @@
+    void initBoardAndValidator(){
+        moveValidator = new MoveValidator();
+        board = new Board();
+    }
```
How come you're using the actual board?  What happened to the test board?

lrharris215 replied:
I was hesitant to make it because all the functions ended up being copies of Board. The only difference is that TestBoard is smaller.

kudukacar replied:
Makes sense.

lrharris215 replied:
Should I take it back out? Already caused an issue w/ the tests bc I had to fix both Board#addDisc and TestBoard#addDisc when I refactored earlier.

kudukacar replied:
I think it's fine to use the actual board when your test board is very similar to the actual board.

lrharris215 committed:
Add: Boardable interface

Added Boardable interface so I can make a test board.

lrharris215 committed:
Fix: Add underscores to const names

lrharris215 committed:
Test: Added TestBoard class

Added the TestBoard class to all tests.

lrharris215 committed:
Test: Added test for Presenter#printError

lrharris215 committed:
Refactor: Moved addDisc logic to Board class

The Player no longer subtracts one from the input column before passing
it to Board#addDisc. Also removed Player#addDisc because it was
unnecessary.

lrharris215 committed:
Removed: TestBoard

TestBoard was just a smaller copy of Board, and wasn't very DRY

lrharris215 committed:
Refactor: Removed Player#takeTurn, add Player#getInput

Removed all placing Discs on Board logic from player, moved to the Game
class. Player now only returns the users input, and Game uses that input
to place the disc on the board.

lrharris215 committed:
Refactor: Moved all const strings to separate Constants class

lrharris215 committed:
Refactor: Created a Player.getDiscColor for readability

pumpkincouture commented on src/main/java/com/connectfour/Presenter.java:
```diff
@@ -1,23 +1,17 @@
-                disc = board.grid[i][j];
+        for(int i = 0; i < board.getGrid().length; i++){
+            for(int j = 0; j < board.getRow(0).length; j++){
+                disc = board.findDisc(i,j);
```
I like that this method is pretty flexible and knows just enough about the inner workings of the board in order to display it :)

kudukacar reviewed (approved):
LGTM!
